proof/crefine/RISCV64/IpcCancel_C.thy

```diff
@@ -2948,198 +3035,256 @@ lemma cancelIPC_ccorres1:
                    (UNIV \<inter> {s. tptr_' s = tcb_ptr_to_ctcb_ptr thread}) []
           (cancelIPC thread) (Call cancelIPC_'proc)"
   apply (cinit lift: tptr_' simp: Let_def cong: call_ignore_cong)
-sorry (* FIXME RT: cancelIPC_ccorres1 *) (*
    apply (rule ccorres_move_c_guard_tcb)
-   apply csymbr
+   apply (rule ccorres_stateAssert)
+   apply (rule ccorres_stateAssert)
    apply (rule getThreadState_ccorres_foo)
-   apply (rule ccorres_symb_exec_r)
-     apply (rule_tac xf'=ret__unsigned_longlong_' in ccorres_abstract, ceqv)
-     apply (rule_tac P="rv' = thread_state_to_tsType rv" in ccorres_gen_asm2)
-     apply wpc
-            \<comment> \<open>BlockedOnReceive\<close>
-            apply (simp add: word_sle_def ccorres_cond_iffs cong: call_ignore_cong)
-            apply (rule ccorres_rhs_assoc)+
-            apply csymbr
-            apply csymbr
-            apply (rule ccorres_pre_getEndpoint)
-            apply (rule ccorres_assert)
-            apply (rule ccorres_symb_exec_r) \<comment> \<open>ptr_get lemmas don't work so well :(\<close>
+   apply (rename_tac threadState)
+   apply csymbr
+   apply csymbr
+   apply (rule ccorres_split_nothrow)
+       apply (rule threadSet_ccorres_lemma2[where P=\<top>])
+        apply vcg
+       apply (clarsimp simp: typ_heap_simps')
+       apply (erule(1) rf_sr_tcb_update_no_queue2, (simp add: typ_heap_simps')+)[1]
+        apply (rule ball_tcb_cte_casesI, simp_all)[1]
+       apply (clarsimp simp: ctcb_relation_def seL4_Fault_lift_NullFault
+                             cfault_rel_def cthread_state_relation_def)
+       apply (case_tac "tcbState tcb", simp_all add: is_cap_fault_def)[1]
+      apply ceqv
+     apply (rule_tac xf'=ret__unsigned_longlong_'
+                 and val="thread_state_to_tsType threadState"
+                 and R="st_tcb_at' ((=) threadState) thread"
+                 and R'=UNIV
+                 in ccorres_symb_exec_r_known_rv)
+        apply (rule conseqPre, vcg)
+        apply (clarsimp simp: st_tcb_at'_def)
+        apply (frule (1) obj_at_cslift_tcb)
+        apply (clarsimp simp: typ_heap_simps ctcb_relation_thread_state_to_tsType)
+       apply ceqv
+      apply wpc
+             \<comment> \<open>BlockedOnReceive\<close>
+             apply (rename_tac oref cg ro)
+             apply (unfold blockedCancelIPC_def)
+             apply (simp add: ccorres_cond_iffs getBlockingObject_BlockedOnReceive_return
+                              return_bind)
+             apply (rule ccorres_rhs_assoc)+
+             apply csymbr
+             apply csymbr
+             apply (rule ccorres_pre_getEndpoint)
+             apply (rule ccorres_assert)
+             apply (rule_tac xf'=ret__unsigned_longlong_'
+                         and val="oref"
+                         and R="st_tcb_at' ((=) (BlockedOnReceive oref cg ro)) thread"
+                         and R'=UNIV
+                         in ccorres_symb_exec_r_known_rv)
```

Comment:
trivial: on these you can save visual space on what's uninteresting by folding the UNIV into:
`in ccorres_symb_exec_r_known_rv[where R'=UNIV]`

indentation: this is sadly my fault, because there's an issue to update the style guide and I haven't done it yet (https://github.com/seL4/l4v/issues/746), but the end of the `in` should align with the end of `and` since this right-aligns the uninteresting glue bits so that the interesting bits can end up left-aligned. Applies to all the big `rule_tac`s in this proof.

Reply:
Done, thanks.